### PR TITLE
fix(membership): invalidate pending site poll cache

### LINKED ISSUE #1314: fix(membership): invalidate wu_membership_meta in pending-site poll handler (cross-worker cache coherence; UM #1308 follow-up)
## Problem

PR #1308 (merged 2026-05-28) adds `wp_cache_delete( $this->get_id(), 'wu_membership_meta' )` inside `Membership::delete_pending_site()` to invalidate the persistent (Docket) cache and the **calling worker's** in-process `WP_Object_Cache`. That fix is necessary but **not sufficient** when the publish loopback and the polling AJAX are handled by different long-lived PHP workers (FrankenPHP worker mode, php-fpm with high `pm.max_requests`, php-pm, RoadRunner, etc.).

**Verified failure on production (`tgc.church`, payment `JJJ3D4416E`, membership 61, blog 90, 2026-05-29 02:08 UTC)**:

1. Browser loads thank-you page. Worker B handles the first poll, calls `$membership->get_pending_site()`, populates `wu_membership_meta` cache for the membership ID in Worker B's in-process `WP_Object_Cache`.
2. Worker A handles the async `publish_pending_site_async()` loopback (`admin-ajax.php?action=wu_publish_pending_site`). After the site is created, `delete_pending_site()` deletes the DB row and runs `wp_cache_delete()` per PR #1308 — clearing Worker A's in-process cache and Docket's disk cache.
3. Browser polls every 5s. Worker affinity keeps routing polls to Worker B. Worker B's in-process cache still holds the pending Site object. `get_pending_site()` returns truthy indefinitely. Thank-you page stuck on "Creating your site…" forever.
4. `wp cache flush` unsticks it (Docket broadcasts flush to all workers).

DB state at the moment of the stuck-poll diagnosis:

```
wp_wu_membershipmeta WHERE wu_membership_id = 61:
  (single row) verified_payment_discount = 1
  (no pending_site row — delete_pending_site DID run)

$membership->get_pending_site()  → still returns a Site object (stale Worker B cache)
$membership->get_pending_site()  → false   ← after `wp cache flush`
```

PR #1308 cannot fix this on its own because there is no IPC mechanism in WP's object-cache contract to broadcast invalidation across PHP worker processes. Each long-lived worker holds an independent in-process mirror of what it has read.

## Proposed fix

Move the cache invalidation into **the read path that polls for completion** — `wp_ajax_wu_check_pending_site_created` (`inc/managers/class-membership-manager.php:205`) — instead of (or in addition to) the write path. That handler runs in the polling worker's request context, so the `wp_cache_delete` it performs invalidates the only worker whose stale cache is actually blocking the user.

The minimal change is one line at the top of the AJAX handler, before it calls `$membership->get_pending_site()`:

```php
// Cross-process invalidation: the publish loopback runs in a different
// long-lived worker, and PR #1308's wp_cache_delete in delete_pending_site()
// only invalidates that worker's in-process cache plus the persistent store.
// We force this poller worker to re-read from the (already-fresh) persistent
// store / DB on every poll, so workers that warmed the meta cache before
// delete_pending_site() fired stop returning stale 'still pending' results.
wp_cache_delete( $membership->get_id(), 'wu_membership_meta' );
```

This is what downstream `tgc.church` already does via a mu-plugin shim (`ai_agent_bust_pending_site_cache`, attached at priority 5 ahead of UM's default 10). Folding the same logic upstream lets every UM consumer benefit without a custom mu-plugin.

## Acceptance criteria

1. In `Membership_Manager::handle_check_pending_site_created()` (or whichever method backs `wp_ajax_wu_check_pending_site_created`), invalidate `wu_membership_meta` for the resolved membership ID **before** calling `get_pending_site()` / `get_sites()`. Use `wp_cache_delete( $membership->get_id(), 'wu_membership_meta' )`.
2. Behaviour when no persistent cache is active: no observable change (`wp_cache_delete` on an empty key is a no-op).
3. Behaviour when Docket / Redis / memcached is active and `delete_pending_site()` has already run on another worker: this poller returns the fresh DB state immediately.
4. Behaviour when `delete_pending_site()` has NOT yet run: this poller still reads `pending_site` correctly — the cache repopulates on the SQL read.
5. Performance: one extra `wp_cache_delete` per poll (one cache key, network-light for Redis/memcached, single `unlink` for Docket). At the 5s polling cadence this is negligible (≤12 cache deletes per minute per active checkout).
6. Add a regression test in `tests/WP_Ultimo/Managers/Membership_Manager_Test.php` (or wherever the existing AJAX handler tests live) that:
   - Warms `wu_membership_meta` cache with a stale pending Site for membership N.
   - Calls the handler directly (or via `do_action( 'wp_ajax_wu_check_pending_site_created' )`).
   - Asserts the response reports `completed` (or whatever signals "no pending site") rather than the stale pending state.

## Files

- `inc/managers/class-membership-manager.php` (one `wp_cache_delete` call near the top of the AJAX handler).
- `tests/WP_Ultimo/Managers/Membership_Manager_Test.php` (one new test method, ~25 LOC).

## Reference pattern

The downstream shim that currently works around this in production:

```php
// tgc.church/site/web/app/mu-plugins/ai-agent-checkout-thankyou.php
add_action( 'wp_ajax_wu_check_pending_site_created', 'ai_agent_bust_pending_site_cache', 5 );
add_action( 'wp_ajax_nopriv_wu_check_pending_site_created', 'ai_agent_bust_pending_site_cache', 5 );

function ai_agent_bust_pending_site_cache(): void {
    if ( ! function_exists( 'wu_get_membership_by_hash' ) ) return;
    $hash = isset( $_REQUEST['membership_hash'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['membership_hash'] ) ) : '';
    if ( empty( $hash ) ) return;
    $membership = wu_get_membership_by_hash( $hash );
    if ( ! $membership ) return;
    wp_cache_delete( $membership->get_id(), 'wu_membership_meta' );
}
```

The upstream version just needs the `wp_cache_delete()` line; membership resolution is already done by UM's handler.

## Non-goals

- Do NOT add cross-process IPC, transient-based broadcasting, or generation counters. This is an unbounded refactor; the polling-side invalidation handles the specific case cheaply.
- Do NOT remove or weaken PR #1308's `wp_cache_delete` in `delete_pending_site()`. That fix is correct for the within-process and persistent-store cases; this PR is an additional fix for the cross-process case.
- Do NOT touch the publish path (`publish_pending_site`, `publish_pending_site_async`). The bug is purely in the read path's cache freshness.

## Verification

After applying the fix on `tgc.church` production:

```bash
# Simulate the stuck state: warm Worker B's cache with stale pending_site.
wp --url=$NETWORK eval '
    wp_cache_delete( 61, "wu_membership_meta" );
    $m = wu_get_membership( 61 );
    $m->get_pending_site();   // warms cache with current DB value
'

# In a separate session, delete the DB row directly without touching cache:
wp --url=$NETWORK db query "DELETE FROM wp_wu_membershipmeta WHERE wu_membership_id=61 AND meta_key='pending_site'"

# Hit the AJAX handler — must report completed, NOT pending.
curl "https://$NETWORK/wp-admin/admin-ajax.php?action=wu_check_pending_site_created&membership_hash=$HASH"
# Expect: {"publish_status":"completed",...}
# Without the fix: {"publish_status":"pending",...}
```

## Notes

Downstream tracking: `tgc.church` retired its `ai_agent_bust_pending_site_cache` shim in commit `16a2fde9` after PR #1308 deployed, then immediately re-restored it in commit `8a6ebbef` when the production failure recurred. The retirement assumed #1308 was sufficient; the failure mode documented here proves it is not. The downstream shim is now back in place with a block comment pointing to this issue.

Filed during post-deploy production verification, 2026-05-29.

---
[aidevops.sh](https://aidevops.sh) v3.20.5 plugin for [OpenCode](https://opencode.ai) v1.15.11 with gpt-5.5 spent 1d 6h and 111,419 tokens on this with the user in an interactive session.

### PR DESCRIPTION
**Pull request by superdav42**
## Summary

- Clear `wu_membership_meta` in the pending-site polling handler before reading `pending_site`.
- Add a regression test that simulates a stale in-process pending-site cache after the DB row has been deleted.

## Testing

- `vendor/bin/phpunit --filter test_check_pending_site_created_invalidates_stale_pending_site_cache`
- `vendor/bin/phpstan analyse inc/managers/class-membership-manager.php tests/WP_Ultimo/Managers/Membership_Manager_Test.php`
- `vendor/bin/phpcs inc/managers/class-membership-manager.php tests/WP_Ultimo/Managers/Membership_Manager_Test.php` (0 errors; 2 pre-existing warnings in the test file)

Resolves #1314

---
[aidevops.sh](https://aidevops.sh) v3.20.5 plugin for [OpenCode](https://opencode.ai) v1.15.12 with gpt-5.5 spent 6m and 119,508 tokens on this as a headless worker.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed an issue where the thank-you page and site creation polling could become stuck. The system now properly clears stale cached data before checking creation status, ensuring users can proceed after their new site is provisioned.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Ultimate-Multisite/ultimate-multisite/pull/1315?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by superdav42** [comment]
MERGE_SUMMARY

Implemented issue #1314 by clearing the membership meta object-cache entry inside the pending-site polling request before the handler reads `pending_site`. This makes the polling worker re-read the already-fresh persistent store or DB when a separate worker completed site publishing and deleted the meta row.

Verification:
- `vendor/bin/phpunit --filter test_check_pending_site_created_invalidates_stale_pending_site_cache`
- `vendor/bin/phpstan analyse inc/managers/class-membership-manager.php tests/WP_Ultimo/Managers/Membership_Manager_Test.php`
- `vendor/bin/phpcs inc/managers/class-membership-manager.php tests/WP_Ultimo/Managers/Membership_Manager_Test.php` (0 errors; 2 pre-existing warnings in the test file)



---
[aidevops.sh](https://aidevops.sh) v3.20.5 plugin for [OpenCode](https://opencode.ai) v1.15.12 with gpt-5.5 spent 7m and 122,472 tokens on this as a headless worker.

**Comment by superdav42** [comment]
BLOCKED: PR #1315 cannot be merged yet because required Cypress checks are failing after a rerun.

What passed:
- PHP Lint
- PHP 8.2 / 8.3 / 8.4 / 8.5 matrix
- Code Quality Checks
- Build Plugin for Testing
- CodeRabbit skipped/passed
- Local targeted regression: `vendor/bin/phpunit --filter test_check_pending_site_created_invalidates_stale_pending_site_cache`
- Local static analysis: `vendor/bin/phpstan analyse inc/managers/class-membership-manager.php tests/WP_Ultimo/Managers/Membership_Manager_Test.php`

Blocking CI evidence:
- `cypress (8.1, chrome)` failed again on rerun job `78430181625`.
- `cypress (8.2, chrome)` failed/cancelled on rerun job `78430181591` after the 8.1 fail-fast.
- Failure is in `tests/e2e/cypress/integration/010-manual-checkout-flow.spec.js`, not the changed pending-site poll path:
  - timed out waiting for `http://localhost:8889/register/` to include `status=done`
  - WP-CLI assertion saw `no-payments` instead of expected `pending`

Triage note: the implementation only clears `wu_membership_meta` before reading `pending_site` in the pending-site polling handler. The repeated Cypress failure is in manual payment creation/checkout completion, so there is not enough evidence to attribute it to this cache invalidation change.



---
[aidevops.sh](https://aidevops.sh) v3.20.5 plugin for [OpenCode](https://opencode.ai) v1.15.12 with gpt-5.5 spent 23m and 217,610 tokens on this as a headless worker.

**Comment by superdav42** [comment]
# Correction: the cross-worker rationale is wrong; the fix is right (different reason)

I authored #1314 and the brief that drove this PR. After empirical testing on the production system that originally exhibited the bug, **my "different long-lived PHP worker" explanation is wrong** and should be retracted from the merged commit message / inline comment / regression-test description. **The code change is still correct, but for a different reason** — sharing the corrected analysis before merge.

## What was wrong with the original rationale

1. **FrankenPHP mode**: the production deploy that exhibited the bug runs FrankenPHP in **classic stateless mode** (`num_threads 2`, no `worker` directive in the Caddyfile). Every HTTP request gets a fresh PHP process state; only OPcache survives across requests. There is no "different long-lived worker" holding a stale in-process `WP_Object_Cache`.
2. **No per-tenant Docket divergence**: I also speculated that `DOCKET_CACHE_CONTENT_PATH` differs across HTTP_HOST contexts (per-subsite vs. network cache). Empirical check on production — across all 11 per-subsite Docket dirs (`/srv/.../shared/tenants/sites/{77,79,80,81,83-90}/cache/cache/docket-cache/`), zero files contain `wu_membership_meta`. The polling AJAX and publish path both land in the network Docket cache. No divergence.

So the original "cross-worker" framing in the issue body, PR description, and inline comment is not what the evidence supports.

## The real cause — write-write race during the publish window

Action Scheduler logs for membership 61 (action 83549, `wu_async_publish_pending_site`):

| Time (UTC) | AS log message |
|---|---|
| 02:05:54 | action created |
| 02:08:02 | action started |
| **02:08:18** | **action complete** (one attempt, no retries; `delete_pending_site()` ran in this tick) |
| 02:08:50 | duplicate scheduling attempt → "action ignored" |
| 02:08:50 → ~02:35 | polling returned `running` for ~27 minutes |
| ~02:35 | `wp cache flush` → next poll returned `completed` |

DB row was demonstrably gone at 02:08:18 (`wp cache flush` only unblocks the read if the row is already absent). Cache held a stale `pending_site` value for 27 minutes. With Docket persistent + stateless FrankenPHP, only a write-write race during the 16-second publish window explains it:

```
T=0    poll N: wp_cache_get($id, 'wu_membership_meta') → miss
T=10ms poll N: SELECT * FROM wp_wu_membershipmeta WHERE wu_membership_id = $id
              → reads pending_site row (still present)
T=15ms publish: DELETE row + wp_cache_delete($id, 'wu_membership_meta')   ← #1308's invalidation
T=20ms poll N: update_meta_cache → wp_cache_set($id, $stale_meta_array)   ← repoisons cache
```

After the race, the cache holds the pending Site object indefinitely. Polling fires every ~5s but every poll hits the (now-stale) cache via `update_meta_cache`'s fast path and never re-reads the DB. The race needs only one poll to land inside the 16-second publish window for the cache to become durably poisoned.

#1308 cannot fix this on its own — it has no mechanism to defer publish's `wp_cache_delete` until after all racing readers complete their `wp_cache_set`. It's still useful (it covers the no-row-to-delete case where WP core's own `delete_metadata()` short-circuits without invalidating), but it's not sufficient.

## Why this PR is the right fix

Adding `wp_cache_delete($membership->get_id(), 'wu_membership_meta')` at the start of `check_pending_site_created()` forces every poll to re-read DB before checking `pending_site`. The maximum stuck window collapses from "indefinite" to "one polling interval (~5s)" — any race-induced stale-cache state gets cleared by the next poll.

The mechanism is **read-path cache invalidation to prevent durably-stuck state after a write-read race**, not "cross-worker IPC". This is what the downstream `tgc.church` mu-plugin shim has been doing in production without recurrence; the issue body's reference pattern is still correct, only its rationale was wrong.

## Suggested wording updates before merge

- **Inline comment** in `class-membership-manager.php` (currently says "the async publish loopback can run in a different long-lived PHP worker"): replace with something like
  > `update_meta_cache` populates `wu_membership_meta` during reads. If a poll races with `delete_pending_site()` and its `wp_cache_set` lands after publish's `wp_cache_delete`, the cache is durably stuck on the stale `pending_site` value. Forcing `wp_cache_delete` on the read path before every poll caps the stuck window to one polling interval.
- **Issue #1314 body** and **PR description**: same correction (the cross-worker / Worker A / Worker B framing).
- **Regression test name** (`test_check_pending_site_created_invalidates_stale_pending_site_cache`): still accurate.
- The test itself: if it currently simulates "different worker holds stale cache", the simulation is fine in practice — write-stale-cache then call handler — but the docblock should be updated to describe the race scenario rather than the worker-affinity scenario.

Happy to push an amend commit if useful; equally happy to leave it to your judgement.

Apologies for the mis-analysis on my end — empirical testing only happened after the PR was open.

Cross-reference: same correction applies to the shim block comment in the downstream `tgc.church` mu-plugin (`ai_agent_bust_pending_site_cache`); I'll update that separately.


---
[aidevops.sh](https://aidevops.sh) v3.20.5 plugin for [OpenCode](https://opencode.ai) v1.15.12 with gpt-5.5 spent 1d 22h and 111,419 tokens on this with the user in an interactive session.
